### PR TITLE
Added proper padding after structs (and matrices).

### LINKED ISSUE #16: structs need padding in std140
the following code

```rust
#[derive(AsStd140)]
struct MainUniform {
    b: Anon,
    c: f32,
}

#[derive(AsStd140)]
struct Anon {
    u: f32,
    v: f32,
}
```

yields:
```
| Offset | Name     | Size |
| ------ | -------- | ---- |
| 0      | _b_align | 0    |
| 0      | b        | 8    |
| 8      | _c_align | 0    |
| 8      | c        | 4    |

[src/main.rs:42] std::mem::size_of::<Std140MainUniform>() = 12
```

I think this is incorrect as struct have 16byte alignment and thus 8byte padding would need to be inserted as _c_align.

https://www.khronos.org/registry/OpenGL/specs/gl/glspec45.core.pdf#page=159
> If the member is a structure, the base alignment of the structure is N, where
N is the largest base alignment value of any of its members, and rounded
up to the base alignment of a vec4. The individual members of this substructure are then assigned offsets by applying this set of rules recursively,
where the base offset of the first member of the sub-structure is equal to the
aligned offset of the structure. The structure may have padding at the end;
the base offset of the member following the sub-structure is rounded up to
the next multiple of the base alignment of the structure.

### LINKED ISSUE #17: arrays of arrays in std430 need padding as well
I think the following layout

```rust
#[derive(AsStd140, AsStd430)]
struct MainUniform {
    e: mint::ColumnMatrix3<f32>,
    f: f32,
}
```

needs to produce the same result in std140 and std430

currently:
```
Std140MainUniform (size 52, alignment 4)
| Offset | Name     | Size |
| ------ | -------- | ---- |
| 0      | _e_align | 0    |
| 0      | e        | 48   |
| 48     | _f_align | 0    |
| 48     | f        | 4    |

[src/main.rs:42] std::mem::size_of::<Std140MainUniform>() = 52

Std430MainUniform (size 40, alignment 4)
| Offset | Name     | Size |
| ------ | -------- | ---- |
| 0      | _e_align | 0    |
| 0      | e        | 36   |
| 36     | _f_align | 0    |
| 36     | f        | 4    |

[src/main.rs:50] std::mem::size_of::<Std430MainUniform>() = 40
```

I think this is the case because e is an array of arrays where each array has an alignment of 16 (rule 3 https://www.khronos.org/registry/OpenGL/specs/gl/glspec45.core.pdf#page=159)

### PR DESCRIPTION
**Pull request by ElectronicRU**
According to GLSL standard, within a struct, sturct and primitive
array members have "padding at end", ie next member essentially has
an alignment no less than the struct or the array itself.

Fixes #16.
Fixes #17.

### CODE REVIEW

File: tests/std140.rs [hunk 1]
```diff
@@ -71,6 +72,33 @@ fn using_vec3_padding() {
...
+#[derive(AsStd140)]
+struct UsingVec3PaddingWithMint {
+    pos: mint::Vector3<f32>,
+    brightness: f32
+}
+
+#[test]
+fn mint_type_produces_same_padding() {
+    let layout_vec: TypeLayoutInfo = <<UsingVec3Padding as AsStd140>::Std140Type as TypeLayout>::type_layout();
+    let layout_mint: TypeLayoutInfo = <<UsingVec3PaddingWithMint as AsStd140>::Std140Type as TypeLayout>::type_layout();
```
Comment (LPGhatguy) [inline tests/std140.rs:84]: Can these types be elided?

File: tests/std140.rs [hunk 2]
```diff
@@ -71,6 +72,33 @@ fn using_vec3_padding() {
...
 }
 
+#[derive(AsStd140)]
+struct UsingVec3PaddingWithMint {
+    pos: mint::Vector3<f32>,
+    brightness: f32
+}
+
+#[test]
+fn mint_type_produces_same_padding() {
```
Comment (LPGhatguy) [inline tests/std140.rs:82]: Can you add a comment talking about what this test is trying to measure and why?

**Review by LPGhatguy — CHANGES_REQUESTED** [review]
Thanks for the PR!

Left a few comments. I think the GitHub workflow has failed due to rustfmt.

File: crevice-derive/src/lib.rs [hunk 3]
```diff
@@ -126,18 +126,34 @@ impl EmitOptions {
...
+                    .take(index)
+                    .next_back()
+                    .map_or(quote!{0usize}, |field|{
+                        let field_ty = &field.ty;
+                        quote! {
+                            ::crevice::internal::pad_at_end(
+                                ::std::mem::size_of::<<#field_ty as #as_trait_path>::#as_trait_assoc>(),
+                                <<#field_ty as #as_trait_path>::#as_trait_assoc as #mod_path::#layout_name>::ALIGNMENT,
+                                <<#field_ty as #as_trait_path>::#as_trait_assoc as #mod_path::#layout_name>::PAD_AT_END
+                            )
```
[the remark this answers is not in the document]
Comment (ElectronicRU) [reply crevice-derive/src/lib.rs:145]: Ah, sorry - I already did that, but didn't push my code.

File: tests/std140.rs (hunk 2 above)
Comment (ElectronicRU) [reply tests/std140.rs]: Nah - I was trying to repro the issue with using the wrong size in offset calculations, but this isn't the right approach.
I just deleted it at this point - didn't manage to get in an example that failed on main.

File: crevice-derive/src/lib.rs [hunk 4]
```diff
@@ -126,18 +126,34 @@ impl EmitOptions {
...
-                                offset += ::std::mem::size_of::<#field_ty>();
+                                offset += ::std::mem::size_of::<<#field_ty as #as_trait_path>::#as_trait_assoc>();
                             }
                         });
 
+                let pad_at_end = fields
+                    .named
+                    .iter()
+                    .take(index)
+                    .next_back()
```
[the remark this answers is not in the document]
Comment (ElectronicRU) [reply crevice-derive/src/lib.rs]: Updated branch.

File: tests/std140.rs (hunk 1 above)
Comment (ElectronicRU) [reply tests/std140.rs]: Outdated - gonna close this now.

**Comment by ElectronicRU** [comment]
Heya! Is it good now?

**Comment by LPGhatguy** [comment]
I put out a call in the Rust game development Discord server for a second set of eyes on this. I think this change is good from my PoV, but I'm not as confident as I think I should be in GLSL layout rules.

**Comment by ElectronicRU** [comment]
Not a proper openGL test, but here's a shader playground link showing the same offsets for the PaddingAtEnd testcase:
http://shader-playground.timjones.io/dfd654fee769dbc00c36bf16cd78c655

**Comment by LPGhatguy** [comment]
This link is very useful! I'd say that I'm confident enough in this change to pull it in and roll with it now.

**Review by LPGhatguy — APPROVED** [review]

**Comment by benmkw** [comment]
I wanted to quickly update the learn-wgpu alignment example but think crevice has still issues. I've unfortunately not yet minimised the problem but the running example can be seen here https://github.com/benmkw/learn-wgpu/commit/b36fef38a3855fd625626e520c065c91f4b454cc and the ground-truth can be seen here https://github.com/LPGhatguy/crevice/pull/9#issuecomment-711140350

I think in the case of std430 its just missing padding at the end for some reason in this more complex case, for std140 it seems to be an issue around `mint::Vector3` and missing padding at the end. 

Sorry for coming up with these examples and not with the solution :/ :D

**Comment by ElectronicRU** [comment]
> I wanted to quickly update the learn-wgpu alignment example but think crevice has still issues. I've unfortunately not yet minimised the problem but the running example can be seen here [benmkw/learn-wgpu@b36fef3](https://github.com/benmkw/learn-wgpu/commit/b36fef38a3855fd625626e520c065c91f4b454cc) and the ground-truth can be seen here [#9 (comment)](https://github.com/LPGhatguy/crevice/pull/9#issuecomment-711140350)
> 
> I think in the case of std430 its just missing padding at the end for some reason in this more complex case, for std140 it seems to be an issue around `mint::Vector3` and missing padding at the end.
> 
> Sorry for coming up with these examples and not with the solution :/ :D

tag @benmkw 

Padding at the end of the whole structure is, interestingly, not mandated by the spec!

(In fact, it specifically does NOT talk about aligning at end of elements, only at starts, which leads to some weird formulations like "array of structures is like laying structures end-to-end", e.g. there are paddings *between* array elements, but *not at the end*, unless the array is itself is inside a structure).

It gets added if you incorporate the struct into another struct (or an array, which isn't supported yet). Also, it doesn't really affect anything - you should be perfectly able to over-allocate your buffers by however much you want. By the spec, you're not required to, though - padding at end is more of an artifact of simplistic OpenGL uniform layout code of yore when every slot was 16 bytes.

Are there problems with inter-member padding? The code you linked is kinda hard to parse at a glance.

@LPGhatguy I'd like in the interest of fairness point out that padding all structures at end for std140 and/or std430 has merit - while slightly space-inefficient, it would get rid of my PAD_AT_END hack, Thoughts?

P.S. I had no idea you're the person making learn-wgpu, gotta say I'm a fan! Helps me slide right into Rust GPU programming.

**Comment by benmkw** [comment]
> Also, it doesn't really affect anything - you should be perfectly able to over-allocate your buffers by however much you want.

I think if an rust array or vec of such elements would get used with e.g. bytemuck, some elements would be misaligned because the rust alignment of the generated structs is not generated correctly because this crate avoids the align attribute for UB reasons IIRC.

> Are there problems with inter-member padding? The code you linked is kinda hard to parse at a glance.

I might have been confused because the layout of an array of 3 floats is in fact not the same as the layout of a vec3, in std140, right? In that case my example is actually not a translation of https://github.com/LPGhatguy/crevice/pull/9#issuecomment-711140350 because we’re still missing arrays. 

Just to clarify, I‘m not the one making learn-wgpu, only tried to make a small contribution regarding the alignment stuff because it comes up ever so often, it can probably be much improved still though as I‘m not using it that much myself.

I found leant wgpu really helpful as well!

**Comment by ElectronicRU** [comment]
> > Also, it doesn't really affect anything - you should be perfectly able to over-allocate your buffers by however much you want.
> 
> I think if an rust array or vec of such elements would get used with e.g. bytemuck, some elements would be misaligned because the rust alignment of the generated structs is not generated correctly because this crate avoids the align attribute for UB reasons IIRC.

If inter-alignment is fine, bytemuck should be fine.

> > Are there problems with inter-member padding? The code you linked is kinda hard to parse at a glance.
> 
> I might have been confused because the layout of an array of 3 floats is in fact not the same as the layout of a vec3, in std140, right? In that case my example is actually not a translation of [#9 (comment)](https://github.com/LPGhatguy/crevice/pull/9#issuecomment-711140350) because we’re still missing arrays.

Yeah - by std140, arrays require padding at the end, but vec3 does not. So, ([f32;3], f32) is 5 words, but (Vec3, f32) is only 4.

I'm working on arrays - they're tricky beasts, the support for generics over them is almost there but not quite.

> Just to clarify, I‘m not the one making learn-wgpu, only tried to make a small contribution regarding the alignment stuff because it comes up ever so often, it can probably be much improved still though as I‘m not using it that much myself.
> 
> I found leant wgpu really helpful as well!
Oh bother, sorry :D

**Comment by benmkw** [comment]
> If inter-alignment is fine, bytemuck should be fine 

if the alignment of the generated struct is less than its size (as its currently frequently the case) I don’t think this is true because in a (rust) array each member will be aligned to its alignment and a successor might thus be at the position of the „optional padding“ of the predecessor. 

Crevice basically expresses alignment by using padding combined with align 4 (or maybe 8?), thus everything which has more than 4 (or maybe 8?) byte alignment needs to make sure that this is respected by introducing the necessary padding (so I think it’s best/ necessary to add the padding at the end, even though it’s optional)

**Comment by ElectronicRU** [comment]
> > If inter-alignment is fine, bytemuck should be fine
> 
> if the alignment of the generated struct is less than its size (as its currently frequently the case) I don’t think this is true because in a (rust) array each member will be aligned to its alignment and a successor might thus be at the position of the „optional padding“ of the predecessor.
> 
> Crevice basically expresses alignment by using padding combined with align 4 (or maybe 8?), thus everything which has more than 4 (or maybe 8?) byte alignment needs to make sure that this is respected by introducing the necessary padding (so I think it’s best/ necessary to add the padding at the end, even though it’s optional)

Yeah - the solution is to hand-craft AsStd140 for arrays with explicit padding, I've actually almost finished it :)

**Comment by benmkw** [comment]
I‘m not talking about a struct with an array member for which one wants to derive std140, but about a plain rust array of the generated structs which one might want to use. If one wants to pass this array of std140 types to the GPU by just casting/ bytemucking it I think misalignment might occur as described the in the last reply.

**Comment by ElectronicRU** [comment]
> I‘m not talking about a struct with an array member for which one wants to derive std140, but about a plain rust array of the generated structs which one might want to use. If one wants to pass this array of std140 types to the GPU by just casting/ bytemucking it I think misalignment might occur as described the in the last reply.

Oh yeah, that's totally fair. An array of Vec3's would be befallen by the same fate, though, the answer is "you probably shouldn't do that".

Once #27 gets finalized, you could call as_std140() on arrays directly, though.

**Comment by benmkw** [comment]
> you could call as_std140() on arrays directly, though

Wouldn't it be more obvious to users if it would 'just work' for arrays (by adding the optional padding to the derived structs)? Not sure if there are any downsides? I imagine if one would want to store the derived types most (space) efficiently, without transferring them to the GPU ... but that seems unlikely?

**Comment by ElectronicRU** [comment]
> > you could call as_std140() on arrays directly, though
> 
> Wouldn't it be more obvious to users if it would 'just work' for arrays (by adding the optional padding to the derived structs)? Not sure if there are any downsides? I imagine if one would want to store the derived types most (space) efficiently, without transferring them to the GPU ... but that seems unlikely?

It'd be more obvious to users, sure, but it'd only work for structs - they're the only type that over-align in both structs and arrays. Vec3s, for example, need special treatment anyway.

I personally think that providing an interface that "almost works" is bad design, but you're welcome to make a PR - maybe LPG is of a different mind!

**Comment by benmkw** [comment]
> It'd be more obvious to users, sure, but it'd only work for structs

Well this crate only works on structs, so you derive a new type for a struct and use that type. Thus it would seem to me like making this type behave as expected in most ways would be useful.
But maybe I’m also missing a use case/ other consideration that exists.

**Comment by ElectronicRU** [comment]
> > It'd be more obvious to users, sure, but it'd only work for structs
> 
> Well this crate only works on structs, so you derive a new type for a struct and use that type. Thus it would seem to me like making this type behave as expected in most ways would be useful.
> But maybe I’m also missing a use case/ other consideration that exists.
Well, vectors, mostly ^_^ it's an edge case, but still.
> That being said, one downside of the „as_std140()“ approach for arrays is probably that it will involve copying the whole array to change the layout/ add the optional padding and then passing that buffer to the GPU.
If you really want to, you could also just directly use YourArrayType::Std140Type!
